scanpy/tools/_score_genes.py

```diff
@@ -134,10 +137,19 @@ def score_genes(
     # Basically we need to compare genes against random genes in a matched
     # interval of expression.
 
-    _adata = adata.raw if use_raw else adata
+    _adata = adata
+    if layer is not None:
+        if use_raw:
+            raise ValueError("Cannot specify `layer` and have `use_raw=True`.")
+        _adata.X = _adata.layers[layer]
+    else:
+        if use_raw and adata.raw is not None:
+            _adata = adata.raw
+
```

Comment:
I believe this line:

```
_adata.X = _adata.layers[layer]
```

Will mutate the input, which we'd like to avoid.

Could this use `scanpy.get._get_obs_rep`?

I *think* this should just be:

```suggestion
        X = _get_obs_rep(adata, use_raw=use_raw, layer=layer)
        if use_raw:
            gene_pool_idx = adata.raw.var_names.get_indexer(gene_pool)
        else:
            gene_pool_idx = adata.var_names.get_indexer(gene_pool)
```

Then use gene_pool_idx to do `X[:, gene+pool_idx]` instead of `_adata_subset`?

Reply:
Maybe??? I copied code from another function but maybe that's also out of date. I'm not sure exactly what those internal functions do but I can try something like that. Is it fine to start by accepting this suggestion or was it more of a sketch?

Reply:
More of a sketch